clients/src/main/java/org/apache/kafka/common/record/FileRecords.java

```diff
@@ -148,6 +154,10 @@ public FileRecords slice(int position, int size) throws IOException {
      * @return the number of bytes written to the underlying file
      */
     public int append(MemoryRecords records) throws IOException {
+        if (size.get() > Integer.MAX_VALUE - records.sizeInBytes())
+            throw new IllegalStateException("Append of size " + records.sizeInBytes() +
```

Comment:
I have a hard time finding the line between these two errors at times. The argument is only invalid given the current size of the file, so wouldn't that mean illegal state? Here is javadoc:
```
/**
 * Signals that a method has been invoked at an illegal or
 * inappropriate time.  In other words, the Java environment or
 * Java application is not in an appropriate state for the requested
 * operation.
 *
 * @author  Jonni Kanerva
 * @since   JDK1.1
 */
public
class IllegalStateException extends RuntimeException {
```

Reply:
IMO, `IllegalStateException` would be appropriate if `FileRecords` was in read-only state for example. But in this case, the expectation is that the caller would only call `append` if `records` fits. So, the caller is passing an illegal argument if that is not the case. I agree that it's a bit of a grey area if `size.get == Integer.MAX_VALUE` though. :)

Reply:
Another question is whether we should be throwing an exception type that the caller can catch to handle it instead of checking before the `append`. That could help the case that @dhruvilshah3 raised.